### PR TITLE
Package exo fixes

### PR DESCRIPTION
**Pull request by josh1593**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -24,7 +24,7 @@
   "rich==13.7.1",
   "tenacity==9.0.0",
   "tqdm==4.66.4",
-  "transformers==4.46.3",
```
Comment (AlexCheema) [inline setup.py:27]: why did you downgrade transformers? is this intentional?

File: exo/download/hf/hf_helpers.py [hunk 2]
```diff
@@ -110,13 +110,14 @@ async def move_models_to_hf(seed_dir: Union[str, Path]):
   source_dir = Path(seed_dir)
   dest_dir = get_hf_home()/"hub"
   await aios.makedirs(dest_dir, exist_ok=True)
-  async for path in source_dir.iterdir():
```
Comment (AlexCheema) [inline exo/download/hf/hf_helpers.py:113]: what was wrong with this?

File: exo/download/hf/hf_helpers.py [hunk 3]
```diff
@@ -110,13 +110,14 @@ async def move_models_to_hf(seed_dir: Union[str, Path]):
...
-    if path.is_dir() and path.startswith("models--"):
-      dest_path = dest_dir / path.name
-      if dest_path.exists():
-        if DEBUG>=1: print(f"skipping moving {dest_path}. File already exists")
-      else:
-        await aios.rename(str(path), str(dest_path))
+  for path in source_dir.iterdir():
+    if path.is_dir() and path.name.startswith("models--"):
+        dest_path = dest_dir / path.name
+        if dest_path.exists():
```
Comment (AlexCheema) [inline exo/download/hf/hf_helpers.py:116]: exists should be made async

**Review by AlexCheema — CHANGES_REQUESTED** [review]

File: setup.py (hunk 1 above)
Comment (josh3519) [reply setup.py]: it was intentional, I was getting errors with the latest version of transformers during compilation, I have found a workaround and no longer uses the 4.43

File: exo/download/hf/hf_helpers.py [hunk 4]
```diff
@@ -110,14 +110,14 @@ async def move_models_to_hf(seed_dir: Union[str, Path]):
   source_dir = Path(seed_dir)
   dest_dir = get_hf_home()/"hub"
   await aios.makedirs(dest_dir, exist_ok=True)
-  async for path in source_dir.iterdir():
-    if path.is_dir() and path.startswith("models--"):
+  async for path in async_iterdir(source_dir):
```
Comment (AlexCheema) [inline exo/download/hf/hf_helpers.py:113]: have you tested this? does `async_iterdir` exist?

File: exo/download/hf/hf_helpers.py [hunk 5]
```diff
@@ -110,14 +110,14 @@ async def move_models_to_hf(seed_dir: Union[str, Path]):
...
   dest_dir = get_hf_home()/"hub"
   await aios.makedirs(dest_dir, exist_ok=True)
-  async for path in source_dir.iterdir():
-    if path.is_dir() and path.startswith("models--"):
+  async for path in async_iterdir(source_dir):  
+    if path.is_dir() and path.name.startswith("models--"):
       dest_path = dest_dir / path.name
-      if dest_path.exists():
-        if DEBUG>=1: print(f"skipping moving {dest_path}. File already exists")
+      if await async_exists(dest_path):
```
Comment (AlexCheema) [inline exo/download/hf/hf_helpers.py:116]: does `async_exists` exist?

File: exo/download/hf/hf_helpers.py (hunk 4 above)
Comment (AlexCheema) [inline exo/download/hf/hf_helpers.py:113]: have you tested this? does async_iterdir exist?

**Review by AlexCheema — CHANGES_REQUESTED** [review]

File: exo/inference/mlx/sharded_utils.py [hunk 6]
```diff
@@ -184,7 +184,7 @@ async def load_shard(
     processor.encode = processor.tokenizer.encode
     return model, processor
   else:
-    tokenizer = await resolve_tokenizer(model_path)
```
Comment (AlexCheema) [inline exo/inference/mlx/sharded_utils.py:187]: Why did you change this?

File: scripts/build_exo.py [hunk 7]
```diff
@@ -24,8 +25,6 @@ def run():
             "--macos-app-mode=gui",
             "--macos-app-version=0.0.1",
             "--macos-signed-app-name=com.exolabs.exo",
-            "--macos-sign-identity=auto",
-            "--macos-sign-notarization",
```
Comment (AlexCheema) [inline scripts/build_exo.py:28]: why were these flags removed? do we need signing?

File: setup.py [hunk 8]
```diff
@@ -8,6 +8,7 @@
   "aiohttp==3.10.11",
   "aiohttp_cors==0.7.0",
   "aiofiles==24.1.0",
+  "anyio==4.6.2",
```
Comment (AlexCheema) [inline setup.py:11]: I don't see why we need to add another dependency. We should be able to do asynchronous file operations with the existing dependencies.

**Review by AlexCheema — CHANGES_REQUESTED** [review]

File: exo/inference/mlx/sharded_utils.py (hunk 6 above)
Comment (AlexCheema) [reply exo/inference/mlx/sharded_utils.py:187]: Not a good reason to change it. Please change it back to `resolve_tokenizer`. If we run into issues with that, we should fix those issues, not revert back to the old version

**Review by AlexCheema — CHANGES_REQUESTED** [review]

File: exo/main.py [hunk 9]
```diff
@@ -211,13 +216,14 @@ async def main():
...
-      await move_models_to_hf(args.models_seed_dir)
+      models_seed_dir = clean_path(args.models_seed_dir)
+      await move_models_to_hf(models_seed_dir)
     except Exception as e:
       print(f"Error moving models to .cache/huggingface: {e}")
 
   # Use a more direct approach to handle signals
   def handle_exit():
-    asyncio.ensure_future(shutdown(signal.SIGTERM, loop))
+    asyncio.ensure_future(shutdown(signal.SIGTERM, loop, node))
```
Comment (AlexCheema) [inline exo/main.py:226]: please be more careful. should be `node.server`.

File: exo/main.py [hunk 10]
```diff
@@ -244,7 +250,7 @@ def run():
   except KeyboardInterrupt:
     print("Received keyboard interrupt. Shutting down...")
   finally:
-    loop.run_until_complete(shutdown(signal.SIGTERM, loop))
+    loop.run_until_complete(shutdown(signal.SIGTERM, loop, node))
```
Comment (AlexCheema) [inline exo/main.py:253]: please be more careful. should be `node.server`.

**Review by AlexCheema — CHANGES_REQUESTED** [review]

File: setup.py [hunk 11]
```diff
@@ -24,7 +24,7 @@
   "rich==13.7.1",
   "tenacity==9.0.0",
   "tqdm==4.66.4",
-  "transformers==4.46.3",
+  "transformers==4.46.3" if (sys.version_info.major==3 and sys.version_info.minor>12) else "transformers==4.43.3" ,
```
Comment (AlexCheema) [inline setup.py:27]: I really don't like this at all. Is there a better solution?

**Review by AlexCheema — CHANGES_REQUESTED** [review]
